src/sqlancer/common/oracle/DQEBase.java

```diff
@@ -0,0 +1,156 @@
+
+    protected final Main.StateLogger logger;
+    protected final MainOptions options;
+    protected final SQLConnection con;
+
+    public DQEBase(S state) {
+        this.state = state;
+        this.con = state.getConnection();
+        this.logger = state.getLogger();
+        this.options = state.getOptions();
+    }
+
+    public abstract String generateSelectStatement(MySQLTables tables, String tableName, String whereClauseStr);
+
+    public abstract String generateUpdateStatement(MySQLTables tables, String tableName, String whereClauseStr);
+
+    public abstract String generateDeleteStatement(MySQLTables tables, String tableName, String whereClauseStr);
+
+    /**
+     * Add auxiliary columns to the database A abstract method, subclasses need to implement it.
+     *
+     * @param table
+     *
+     * @throws SQLException
+     */
+    public abstract void addAuxiliaryColumns(AbstractRelationalTable<?, ?, ?> table) throws SQLException;
+
+    public void dropAuxiliaryColumns(AbstractRelationalTable<?, ?, ?> table) throws SQLException {
+        String tableName = table.getName();
+        String dropColumnRowId = String.format("ALTER TABLE %s DROP COLUMN %s", tableName, COLUMN_ROWID);
+        new SQLQueryAdapter(dropColumnRowId).execute(state);
+        String dropColumnUpdated = String.format("ALTER TABLE %s DROP COLUMN %s", tableName, COLUMN_UPDATED);
+        new SQLQueryAdapter(dropColumnUpdated).execute(state);
+    }
+
+    // This interface is to record Error code
+    public interface UpdateErrorCodes {
+
+    }
+
+    public interface ErrorCodeStrategy {
+        Set<Integer> getUpdateSpecificErrorCodes();
+
+        Set<Integer> getDeleteSpecificErrorCodes();
+
+    }
+
+    /**
+     * The core idea of DQE is that the SELECT, UPDATE and DELETE queries with the same predicate φ should access the
+     * same rows. If these queries access different rows, DQE reveals a potential logic bug in the target DBMS.
+     */
+    public static class SQLQueryResult {
+
+        private final Map<AbstractRelationalTable<?, ?, ?>, Set<String>> accessedRows; // Table name with respect rows
+        private final List<SQLQueryError> queryErrors;
+
+        public SQLQueryResult(Map<AbstractRelationalTable<?, ?, ?>, Set<String>> accessedRows,
+                List<SQLQueryError> queryErrors) {
+            this.accessedRows = accessedRows;
+            this.queryErrors = queryErrors;
+        }
+
+        public Map<AbstractRelationalTable<?, ?, ?>, Set<String>> getAccessedRows() {
+            return accessedRows;
+        }
+
+        public List<SQLQueryError> getQueryErrors() {
+            return queryErrors;
+        }
+
+        public boolean hasEmptyErrors() {
+            return queryErrors.isEmpty();
+        }
+
+        public boolean hasErrors() {
+            return !hasEmptyErrors();
+        }
+
+        public boolean hasSameErrors(SQLQueryResult that) {
```

Comment:
Since this method invokes `equals` on all the elements, I think this method should probably override `equals` as well?